packages/keyring-controller/src/KeyringController.ts

```
@@ -2114,6 +2201,9 @@ export class KeyringController extends BaseController<
           state.encryptionKey = updatedState.encryptionKey;
           state.encryptionSalt = JSON.parse(updatedState.vault as string).salt;
         }
+        if (updatedKeyrings.length < state.keyringsMetadata.length) {
+          state.keyringsMetadata = state.keyringsMetadata.slice(0, -1 * (state.keyringsMetadata.length - updatedKeyrings.length));
+        }
```

Hmm, this feels like an unexpected situation where I'm not sure it would be safe to 'just delete the last ones'. Maybe we should just throw an error here?

---

Hmm but if we throw an error here, the user would not be able to load the extension.

---

That's a good point, but I'm also thinking that when a consumer initialises this `KeyringController` version coming from an older version we'll have an empty `keyringMetadata` array, which would create several issues in how these are matched with the related keyring

Perhaps we should have checks in place for both scenarios, and a strategy to realign them for when it happens. 

The issue that I see is that if other controllers keep track of these IDs (which could be since keyringMetadata is publicly accessible), they could end up keeping track of unreferenced IDs if we remove / change them

---

Also, what happens if a keyring from the middle of the array is removed?

---

For the clients, there will be a migration to generate the keyringMetadata for all keyrings. 

I've added checks to clear the metadata of a keyring if its [removed](https://github.com/MetaMask/core/pull/5112/commits/0e8b42a8b17dab4455bbd35f0b9d60e2b859f75e#diff-cacd62f731d7f6947fe841f6ba3441c7a16f764b699801b0c212e0e6c15879bfR2370-R2397).

We also probably need a new event to notify controllers of metadata changes, like `KeyringController:keyringMetadataChanged`.

---

The thing i'm not sure how to handle is, if there is a mismatch between the number of keyrings and keyringsMetadata. I don't think there is a reliable way to match the id to the keyring instance, since we are currently only using the index.

---

> For the clients, there will be a migration to generate the keyringMetadata for all keyrings.

About this, I don't think there is a way to get a list of available keyrings or to manipulate the keyrings array before the vault is injected into KeyringController via constructor and `submitPassword` is called with the password typed by the user. So KeyringController will have to deal with the possibility of receiving keyrings with no ID, even just to support older states

---

> > For the clients, there will be a migration to generate the keyringMetadata for all keyrings.
> 
> About this, I don't think there is a way to get a list of available keyrings or to manipulate the keyrings array before the vault is injected into KeyringController via constructor and `submitPassword` is called with the password typed by the user. So KeyringController will have to deal with the possibility of receiving keyrings with no ID, even just to support older states

I used the number of keyring types thats available in the accounts controller to determine the number of `KeyringMetadata` to be created. There should only be one unique instances of each keyring because the multi srp feature isn't live. But youre right, if we want to use `KeyringController.state.keyrings` it would have to be after `submitPassword` is called. I think there is a need for this type of migration tool.

---

> I used the number of keyring types thats available in the accounts controller to determine the number of KeyringMetadata to be created.

That would indeed work but not if the user has multiple imported accounts, since each of them will have its own SimpleKeyring instance